sensors/sensor.py

```diff
@@ -45,15 +43,39 @@ def occupied_status(self):
         raise NotImplementedError('This method must be implemented in the subclasses')
 
     def open_connection(self):
-        return self.__piGPIO.i2c_open(self.__bus, self.__address) # open I2C device with respective bus and address
+        return self.__pi.i2c_open(self.__bus, self.__address)
 
     def close_connection(self):
-        return self.__piGPIO.i2c_close(self.__handle)
+        return self.__pi.i2c_close(self.__handle)
 
     def read_raw_data(self):
-        return self.__piGPIO.i2c_read_device(self.__handle, self.__buffer_length)
+        return self.__pi.i2c_read_device(self.__handle, self.__buffer_length)
 
     def write_raw_data(self, handle, start_command):
-        return self.__piGPIO.i2c_write_device(self.__handle, [start_command])
+        return self.__pi.i2c_write_device(handle, start_command)
+
+    @staticmethod
+    def __get_pi_gpio():
+        """
+        Creates one instance of pigpio.pi() so that we are not establishing
+        a connection for each sensor. We only need one connection across all sensors
+        to the pi. Hence, we treat __pi.gpio.instance as a singleton instance, and
+        use this method to obtain the singleton instance.
+
+        :return: pigpio.pi instance
+        """
+        if Sensor.__pi_gpio_instance is None:
+            print("Attempting to connect to Raspberry Pi GPIO...")
+            Sensor.__pi_gpio_instance = pigpio.pi('pi3.local')
+
+            if not Sensor.__pi_gpio_instance.connected:
+                print("Failed to connect to pi GPIO daemon")
+                if CONFIG.debug_config.enabled:
+                    print("Using GPIO stub.")
+                    Sensor.__pi_gpio_instance = PiStub('pi3.local')
+                else:
+                    exit(-1)
 
+            print("Successfully Connected! (Version: " + str(Sensor.__pi_gpio_instance.get_pigpio_version()) + ")\n")
```

Comment:
`String#format`

Reply:
I think it would be good practice to have a `getInstance` method for `__pi_gpio_instance`

Reply:
`f"{var}"` is even cleaner

Reply:
I always forget that python has this, it's such a nice convenience!